src/enumerable.cr

```diff
@@ -558,6 +558,25 @@ module Enumerable(T)
     raise Enumerable::NotFoundError.new
   end
 
+  # Yields each value until the first truthy block result and returns that result.
+  #
+  # Accepts an optional parameter `if_none`, to set what gets returned if
+  # no element is found (defaults to `nil`).
+  #
+  # ```
+  # [1, 2, 3, 4].find_value { |i| i > 2 }     # => true
+  # [1, 2, 3, 4].find_value { |i| i > 8 }     # => nil
+  # [1, 2, 3, 4].find_value(-1) { |i| i > 8 } # => -1
+  # ```
+  def find_value(if_none = nil, & : T ->)
```

Comment:
I think it would be of some benefit to have the method strictly typed rather than leaving the compiler to infer everything. We had a discussion in the Discord server and concluded it would mean having an overload specifically for a `nil`/no-default case, but I think that would be better overall:

```crystal
def find_value(if_none : U, & : T -> V) : V forall U, V
def find_value(& : T -> U) : U? forall U
```

Reply:
I'm a fan of explicit types, but this is still type inference. What benefit are you seeing here that I'm not seeing?

Reply:
I find that the explicitness of method signatures are more useful even if there is no real difference between them. It's also clear what the return type of the method is as both I and another Crystal user initially misinterpreted the return type as being the value of the enumerable type (i.e. `T`). When taking into account [`Enumerable#find`](https://crystal-lang.org/api/1.13.1/Enumerable.html#find%28if_none%3Dnil%2C%26%3AT-%3E%29-instance-method) which has an identical signature, it makes sense to be explicit here to reduce confusion.

Reply:
I agree that it's more explicit, but I don't know if I agree that it reduces confusion. To my eyes, it just looks like a jumble of type placeholders.

Does the doc comment provide insufficient disambiguation between it and `find`?

<img width="590" alt="screenshot of the documentation for the find_value method. It reads: Yields each value until the first truthy block result and returns that result." src="https://github.com/user-attachments/assets/40f1f4bc-45b0-49e8-a33b-fd94f2b39ad1">

Reply:
I think type restrictions are always helpful as they document the expectations of input and ouput types. Even if it's a bit complicated to express.
IMO we should ideally always write down all type restrictions as part of the API documentation.

As a comment on the suggested format, different names `T -> V` and `T -> U` are confusing. Both proc types should use the same name for their output type.

Reply:
> Even if it's a bit complicated to express.

My argument isn't against it being difficult to express. It's about it being more difficult to read.

The code as it currently exists in this branch is easier to read and aligns with existing conventions.

Reply:
In my opinion, public API methods should always have as many and as detaillled type restrictions as possible.
If we _can_ type it, we _should_ type it. This has not always happened in the past, and might always be the case in the future. But I'd prefer it that way.

I don't find it helpful to leave out relevant type information for the sake of readability.
Nobody _needs_ to read the type restrictions if they don't care about them. But if you care about them, they should be available.

That being said, I'm happy to accept this PR without the additional type information. We can add it in a follow-up (or not).

Reply:
> In my opinion, public API methods should always have as many and as detaillled type restrictions as possible. … I don't find it helpful to leave out relevant type information for the sake of readability.

Since we have type inference in Crystal, what purpose do type annotations (beyond those needed to satisfy the compiler) serve, if not clarity for the reader?

For the record, I *am* a fan of type annotations in Crystal. 😄 [I use them most of the time](https://forum.crystal-lang.org/t/incremental-compilation-exploration/5224/22?u=jgaskins) when they improve clarity. But I think there's a point where they start providing negative value and I also think this is one of those times.

Reply:
The purpose of type annotations is to provide information to the user.
No annotations means less information, so it's less useful.

There might be a point where the types are too complex to express with the type grammar. But I don't think this is too much complexity:
```cr
def find_value(if_none : V, & : T -> U) : U | V forall U, V
def find_value(& : T -> U) : U? forall U
```

Reply:
I see that many other methods in `Enumerable` just use `& : T ->` while some others use an explicit `forall`... Maybe this whole thread is just circumstantial to the feature, and we could continue this discussion on a follow up normalization of `Enumerable` (and other) type signatures?

That would allow us to avoid delay merging this PR any longer :slightly_smiling_face: